File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        if (k >= nums.size()) return nums.size();

        int numZeros = 0, left = 0, right = 0, maxL = 0;

        while (right < nums.size()) {
            //increment numZeros
            if (nums[right] == 0) {
                numZeros++;
            }

            //once numZeros is greater than k
            if (numZeros > k) {
                //increment left to past next zero
                if (nums[left] == 0) {
                    numZeros--;
                }
                left++;
            }

            //update max
             maxL = max(right - left + 1, maxL);

            right++;
        }

        return maxL;
    } //runtime : O(N) , space : O(1)
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        left = 0
        for right in range(len(nums)):
            print(k, nums[right])
            k -= (1 - nums[right])
            if k < 0:
                if nums[left] == 0:
                    k += 1
                left += 1
        return right - left + 1
            
